### PR TITLE
RPC Implementation using multiple relays to enable async communication

### PR DESCRIPTION
**Pull request by L3tum**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ❌ 
| New feature?  | ✔️ TODO please update "Other Features" section in CHANGELOG.md file
| Issues        | https://github.com/orgs/roadrunner-server/discussions/1850 
| Docs PR       | spiral/docs#...  TODO?

Implements an "Async" RPC Interface and Implementation using multiple relays to effectively offer non-blocking IO in regards to the Roadrunner communication.

As discussed in the linked discussion, there's a (small but signifcant) delay in the RPC communication to Roadrunner that *can* have an impact on a service. In my case the service had a nominal response time of ~2ms, went up to 3ms with metric collection and also had a 2ms `kernel.terminate` listener to collect further metrics, so all in all was busy for ~5ms with only ~2ms spent on the actual request->response cycle.

As a fix I propose an "async" implementation of the RPC interface. I don't really like calling it async, because it doesn't offer the same versatility as a proper async implementation, but it changes the control flow and thus qualifies for that name, I guess.
The interface has 3 significant new methods as well as a few supporting changes in other code:
- callIgnoreResponse => Calls a Roadrunner method but doesn't wait for the response and ignores it if it's received. This is intended for "best-effort" areas such as Metrics collection, where it is generally accepted if some Metrics get lost on the way
- callAsync => Calls a Roadrunner method and tracks the response rather than wait for it. The response can then be received with...
- getResponse => Gets and if needed waits for a response from a previous call to callAsync. The request/response is tracked with the already used $seq. The implementation keeps the response in the socket buffer as long as possible until it's needed to reduce unnecessary delays
- The supporting methods `hasResponse` and `hasAnyResponse` just check if a response has been received

**Results**
I've measured the impact in a testcase using the same repro used in the discussion. The results are the following:
- Sync Metrics Call: 1.228ms
- AlwaysIgnoreResponseRPC (from the repro) Call: 0.1366ms
- MultiRPC Call (this change): 0.1399ms
As you can see the "good" implementation (this change) is within the normal deviation of the always-ignore implementation and offers an improvement ~10x performance. I've been using a (slightly modified) implementation of this in the aforementioned production service and could cut down the time spent collecting metrics from a total of 3ms to just 0,2ms, with a min/max of respectively 0.1ms and 0.3ms.

I've also added a change (PR following) in roadrunner-kv that uses the `callAsync` method to implement a `setAsync` function. The results so far are promising. With a follow-up check on whether the `kv.Set` call was successful, the async implementation is significantly faster:

- "SyncKVImpactCacheStartup": "56.2324ms"
- "SyncKVImpactNewItem": "0.5328ms"
- "SyncKVImpactExistingItemSameValue: "0.2113ms"
- "SyncKVImpactExistingItemNewValue": "0.2006ms"
- "AsyncKVImpactCacheStartup": "21.7821ms"
- "AsyncKVImpactNewItem": "0.1146ms"

Full results for the Metrics calls:
````json
{
    "Base": {
        "Max": "0.2063ms",
        "Min": "0.0954ms",
        "Avg": "0.1228ms",
        "Total": "1.228ms"
    },
    "AlwaysIgnoreResponseRPC": {
        "Max": "0.0537ms",
        "Min": "0.0069ms",
        "Avg": "0.01366ms",
        "Total": "0.1366ms"
    },
    "MultiRPC": {
        "Max": "0.0714ms",
        "Min": "0.0061ms",
        "Avg": "0.01399ms",
        "Total": "0.1399ms"
    }
}
````

**Considerations**
- The whole implementation is "opt-in", which can also be seen in my two follow-up PRs ([#1](https://github.com/roadrunner-php/metrics/pull/13) and [#2](https://github.com/roadrunner-php/kv/pull/36)). I didn't want to force it in case someone depends on the existing implementation. It makes some of the code look a bit messy due to having to support both code paths but I think it's the better solution. I'm open to change it though if you disagree.
- I had to "open" two previously private fields in the Relay implementations to make more effective use of `socket_select` and `stream_select`. I've commented it as well as marked them with `@internal` but it's definitely not ideal.
- ~The overflow response buffer currently has no upper limit and thus it's somewhat easy to get a memory leak here. I could add a check in `callAsync` to see if the response buffer is above a certain limit (1000?) and if so flush it. I'll probably do so when adding the tests.~ Implemented a flush at 1000 entries (realistically the most a normal application should ever see is maybe ~50) -- I've upped this to 10_000 entries because in testing in particular Doctrine caches get hammered extremely hard and 1_000 is not enough.

**TODO**
- ~I have glanced at the tests in this repo and will try to implement a few for this change if you've given me the go-ahead with the current implementation.~ Added a number of tests
- Add a Changelog.md entry
- Add documentation for this (?)

## Summary by CodeRabbit

- **New Features**
    - Introduced `ConnectedRelayInterface` for explicit connection management.
    - Added `MultiRelayHelper` for relay selection and connection status checks.
    - Enhanced RPC functionality with `AbstractRPC`, `AsyncRPCInterface`, and `MultiRPC` for synchronous and asynchronous calls.
    - Implemented `SocketRelay` and `StreamRelay` changes for improved cloning and visibility adjustments.
- **Bug Fixes**
    - Corrected a typo in `RPCInterface` method description.
- **Refactor**
    - Updated `RPC` class to extend `AbstractRPC` with changes in constructor and method adjustments.
- **Tests**
    - Added comprehensive test suites for new RPC and relay functionalities, including exception handling and multi-relay support.
- **Documentation**
    - Improved internal documentation for public properties in `SocketRelay` and `StreamRelay`.
- **Chores**
    - Renamed test class for consistency in TCP RPC tests.
- **New Features in Testing**
    - Introduced a `SleepEcho` function in test server for delayed response simulation.

### CODE REVIEW

**Comment by L3tum** [comment]
Hey @rustatian , @wolfy-j , @msmakouz , @roxblnfk , took a bit of maneuvering to get the commits across but this is the new PR with the latest changes in place.
Interestingly the phpunit test only fails on github, running it locally makes it pass. It's probably due to using STDIN/STDOUT and that not being present in github actions. I'll use a temp stream for that test as well

File: src/RPC/MultiRPC.php [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+                    )
+                );
+            }
+        }
+
+        self::$freeRelays = $relays;
+        self::$occupiedRelays = self::$seqToRelayMap = self::$asyncResponseBuffer = [];
+        $this->asyncBufferThreshold = $asyncBufferThreshold;
+        parent::__construct($codec);
+    }
```
[the remark this answers is not in the document]
Comment (L3tum) [reply src/RPC/MultiRPC.php]: I've added a comment explaining the decision.

**Comment by L3tum** [comment]
I've pushed the aforementioned fix, alongside a couple of additional tests and more checks in MultiRPC to make sure it's running correctly. I've also added a lengthy comment explaining the decision behind using static properties rather than instance variables.

Tests should pass now. Let me know if anything is amiss :)

**Comment by rustatian** [comment]
@L3tum 👋 
You may use https://nektosact.com/ to test your GitHub actions locally 😃

**Comment by rustatian** [comment]
@L3tum I'm not sure, is that really needed to include a Go code in the PHP tests? Might be it would be better to use RR binary?

**Comment by L3tum** [comment]
@rustatian I'm not well versed in the out-of-the-box capabilities of RR for whether it'd be possible to replace all the tests. In the end it would probably make sense to at least add integration tests to all of the PHP SDK packages. But I'll defer that decision to the maintainers of it :)

**Review by msmakouz — APPROVED** [review]

**Comment by msmakouz** [comment]
@L3tum Thanks for the work you've done!

**Comment by L3tum** [comment]
@msmakouz No problem, thanks for reviewing&merging it!

**Pull request merged**